diff --git a/node_modules/@shapediver/viewer/dist/implementation/viewport/ViewportApi.js b/node_modules/@shapediver/viewer/dist/implementation/viewport/ViewportApi.js
index ddfa16d..31dfdda 100644
--- a/node_modules/@shapediver/viewer/dist/implementation/viewport/ViewportApi.js
+++ b/node_modules/@shapediver/viewer/dist/implementation/viewport/ViewportApi.js
@@ -747,13 +747,13 @@ class ViewportApi {
             this.update('createArSessionLink.end');
             const response = yield arSessionEngine.uploadGLTF(new Blob([blob], { type: 'application/octet-stream' }), sdk_geometry_api_sdk_v2_1.ShapeDiverRequestGltfUploadQueryConversion.SCENE);
             const backends = {
-                "sddev3": "https://sddev3.eu-central-1.shapediver.com",
-                "sddev2": "https://sddev2.eu-central-1.shapediver.com",
-                "sddev": "https://sddev.eu-central-1.shapediver.com",
-                "sdtest": "https://sdtest.us-east-1.shapediver.com",
-                "sdeuc1": "https://sdeuc1.eu-central-1.shapediver.com",
-                "sdr7euc1": "https://sdr7euc1.eu-central-1.shapediver.com",
-                "sduse1": "https://model-view.shapediver.com",
+                "sddev3": "http://192.168.0.10:8081/",
+                "sddev2": "http://192.168.0.10:8081/",
+                "sddev": "http://192.168.0.10:8081/",
+                "sdtest": "http://192.168.0.10:8081/",
+                "sdeuc1": "http://192.168.0.10:8081/",
+                "sdr7euc1": "http://192.168.0.10:8081/",
+                "sduse1": "http://192.168.0.10:8081/",
             };
             let backendIdentifier = Object.keys(backends).find((key) => backends[key] === arSessionEngine.modelViewUrl);
             if (!backendIdentifier) {
diff --git a/node_modules/@shapediver/viewer/src/implementation/viewport/ViewportApi.ts b/node_modules/@shapediver/viewer/src/implementation/viewport/ViewportApi.ts
index 2dc8bf1..dc77caf 100644
--- a/node_modules/@shapediver/viewer/src/implementation/viewport/ViewportApi.ts
+++ b/node_modules/@shapediver/viewer/src/implementation/viewport/ViewportApi.ts
@@ -761,7 +761,7 @@ export class ViewportApi implements IViewportApi {
         return this.#renderingEngine.mouseEventToRay(event);
     }
 
-    public raytraceScene(origin: vec3, direction: vec3, root?: ITreeNode): { distance: number, node: ITreeNode, data: IGeometryData; }[] {
+    public raytraceScene(origin: vec3, direction: vec3, root?: ITreeNode): { distance: number, node: ITreeNode, data?: IGeometryData; }[] {
         const scope = 'raytraceScene';
         this.#inputValidator.validateAndError(`ViewportApi.${scope}`, origin, 'vec3');
         this.#inputValidator.validateAndError(`ViewportApi.${scope}`, direction, 'vec3');
@@ -883,13 +883,13 @@ export class ViewportApi implements IViewportApi {
         const backends: {
             [key: string]: string
         } = {
-            "sddev3": "https://sddev3.eu-central-1.shapediver.com",
-            "sddev2": "https://sddev2.eu-central-1.shapediver.com",
-            "sddev": "https://sddev.eu-central-1.shapediver.com",
-            "sdtest": "https://sdtest.us-east-1.shapediver.com",
-            "sdeuc1": "https://sdeuc1.eu-central-1.shapediver.com",
-            "sdr7euc1": "https://sdr7euc1.eu-central-1.shapediver.com",
-            "sduse1": "https://model-view.shapediver.com",
+            "sddev3": "http://192.168.0.10:8081/",
+            "sddev2": "http://192.168.0.10:8081/",
+            "sddev": "http://192.168.0.10:8081/",
+            "sdtest": "http://192.168.0.10:8081/",
+            "sdeuc1": "http://192.168.0.10:8081/",
+            "sdr7euc1": "http://192.168.0.10:8081/",
+            "sduse1": "http://192.168.0.10:8081/",
         }
 
         let backendIdentifier = Object.keys(backends).find((key: string) => backends[key] === arSessionEngine.modelViewUrl);